lib/src/message/message_item_bloc.dart

```diff
@@ -47,112 +47,186 @@ import 'package:bloc/bloc.dart';
 import 'package:delta_chat_core/delta_chat_core.dart';
 import 'package:ox_coi/src/data/repository.dart';
 import 'package:ox_coi/src/data/repository_manager.dart';
+import 'package:ox_coi/src/data/repository_stream_handler.dart';
+import 'package:ox_coi/src/l10n/l.dart';
+import 'package:ox_coi/src/l10n/l10n.dart';
 import 'package:ox_coi/src/message/message_item_event_state.dart';
 import 'package:ox_coi/src/ui/color.dart';
 import 'package:ox_coi/src/utils/date.dart';
 
 class MessageItemBloc extends Bloc<MessageItemEvent, MessageItemState> {
   Repository<Contact> _contactRepository = RepositoryManager.get(RepositoryType.contact);
+  RepositoryMultiEventStreamHandler _repositoryStreamHandler;
   Repository<ChatMsg> _messageListRepository;
   int _messageId;
-  int _nextMessageId;
   int _contactId;
-  bool _showContact;
+  bool _listenersRegistered = false;
 
   @override
   MessageItemState get initialState => MessageItemStateInitial();
 
   @override
   Stream<MessageItemState> mapEventToState(MessageItemEvent event) async* {
-    if (event is RequestMessage) {
-      try {
-        var chatId = event.chatId;
-        _messageListRepository = RepositoryManager.get(RepositoryType.chatMessage, chatId);
-        _messageId = event.messageId;
-        _nextMessageId = event.nextMessageId;
-        _showContact = event.isGroupChat || isInvite(chatId);
-        if (_showContact) {
-          await _setupContact();
-        }
-        if (_hasNextMessage()) {
-          await _setupNextMessage();
-        }
-        await _setupMessage();
-        add(MessageLoaded());
-      } catch (error) {
-        yield MessageItemStateFailure(error: error.toString());
+    if (event is LoadMessage) {
+      yield* _loadMessage(event);
+    } else if (event is DeleteMessage) {
+      yield* _deleteMessages(event.id);
+    } else if (event is FlagUnflagMessage) {
+      yield* _flagUnflagMessage(event.id);
+    } else if (event is MessageUpdated) {
+      yield MessageItemStateSuccess(messageStateData: event.messageStateData);
+    }
+  }
+
+  @override
+  void close() {
+    unregisterListeners();
+    super.close();
+  }
+
+  Stream<MessageItemState> _loadMessage(LoadMessage event) async* {
+    try {
+      var chatId = event.chatId;
+      _messageListRepository = RepositoryManager.get(RepositoryType.chatMessage, chatId);
+      _messageId = event.messageId;
+      var nextMessageId = event.nextMessageId;
+      var isGroup = event.isGroupChat;
+      var showContact = isGroup || chatId == Chat.typeInvite;
+      if (showContact) {
+        await _setupContact();
       }
-    } else if (event is MessageLoaded) {
-      bool showTime = await _getShowTime();
-      bool encryptionStatusChanged = await _getEncryptionStatusChanged();
+      if (nextMessageId != null) {
+        await _setupNextMessage(nextMessageId);
```

Comment:
Just for clarification: What is that `nextMessageId` for?

Reply:
We have two cases where the UI depends on the difference of two messages:

- Encryption state changed between two messages
- Date time changed between two messages (difference of at least one minute)

To check those states we hand over the message id and the next message id. So the BloC can calculate differences and give the UI enough information if specific states should be shown.